src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -0,0 +1,666 @@
+#include "db_sqlite.h"
+
+#include <sodium.h>
+#include <SQLiteCpp/SQLiteCpp.h>
+#include <sqlite3.h>
+
+#include <string>
+#include <iostream>
+#include <cassert>
+
+#include "cryptonote_core/blockchain.h"
+#include "cryptonote_core/service_node_list.h"
+#include "common/string_util.h"
+#include "cryptonote_basic/hardfork.h"
+
+#undef OXEN_DEFAULT_LOG_CATEGORY
+#define OXEN_DEFAULT_LOG_CATEGORY "blockchain.db.sqlite"
+
+namespace cryptonote
+{
+
+BlockchainSQLite::BlockchainSQLite(cryptonote::network_type nettype, fs::path db_path)
+  : db::Database(db_path, ""), m_nettype(nettype), filename(db_path.u8string())
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+  height = 0;
+
+  if (!db.tableExists("batched_payments_accrued") || !db.tableExists("batched_payments_raw") || !db.tableExists("batch_db_info")) {
+    create_schema();
+  }
+
+  SQLite::Statement st{db, "SELECT height FROM batch_db_info"};
+  while (st.executeStep()) {
+    this->height = st.getColumn(0).getInt64();
+  }
+}
+
+void BlockchainSQLite::create_schema() {
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+
+	db.exec(R"(
+CREATE TABLE batched_payments_accrued (
+    address VARCHAR NOT NULL,
+    amount BIGINT NOT NULL,
+    PRIMARY KEY(address),
+    CHECK(amount >= 0)
+);
+
+CREATE TRIGGER batch_payments_delete_empty
+AFTER UPDATE ON batched_payments_accrued
+FOR EACH ROW WHEN NEW.amount = 0
+BEGIN
+    DELETE FROM batched_payments_accrued WHERE address = NEW.address;
+END;
+
+CREATE TABLE batched_payments_raw (
+    address VARCHAR NOT NULL,
+    amount BIGINT NOT NULL,
+    height_paid BIGINT,
+    PRIMARY KEY(address, height_paid),
+    CHECK(amount >= 0)
+);
+
+CREATE INDEX batched_payments_raw_height_idx ON batched_payments_raw (height_paid);
+
+CREATE TABLE batch_db_info (
+    height BIGINT NOT NULL
+);
+
+INSERT INTO batch_db_info (height) VALUES (0);
+
+CREATE TRIGGER batch_payments_prune
+AFTER UPDATE ON batch_db_info
+FOR EACH ROW
+BEGIN
+    DELETE FROM batched_payments_raw WHERE height_paid < (NEW.height - 10000);
+END;
+
+CREATE VIEW batched_payments_paid AS SELECT * FROM batched_payments_raw;
+
+CREATE TRIGGER make_payment INSTEAD OF INSERT ON batched_payments_paid
+FOR EACH ROW
+BEGIN
+    UPDATE batched_payments_accrued SET amount = (amount - NEW.amount) WHERE address = NEW.address;
+    SELECT RAISE(ABORT, 'Address not found') WHERE changes() = 0;
+    INSERT INTO batched_payments_raw (address, amount, height_paid) VALUES (NEW.address, NEW.amount, NEW.height_paid);
+END;
+
+CREATE TRIGGER rollback_payment INSTEAD OF DELETE ON batched_payments_paid
+FOR EACH ROW
+BEGIN
+    DELETE FROM batched_payments_raw WHERE address = OLD.address AND height_paid = OLD.height_paid;
+    INSERT INTO batched_payments_accrued
+        (address, amount) VALUES (OLD.address, OLD.amount)
+        ON CONFLICT (address) DO UPDATE SET amount = (amount + excluded.amount);
+END;
+	)");
+
+	MDEBUG("Database setup complete");
+}
+
+void BlockchainSQLite::clear_database()
```

Comment:
I think this should be renamed to `reset_database()` because it's doing a bit more than just clearing it.